<a href="https://colab.research.google.com/github/SvetlanaTsim/ml_in_business/blob/main/course_project/25_01_gb_ml_course_project_request.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Строим запрос к нашему сервису

In [ ]:
import requests

import urllib.request
import json 

In [ ]:
# Пример данных
data  = {'gender': 'Male',
 'MultipleLines': 'No',
 'InternetService': 'DSL',
 'OnlineSecurity': 'No',
 'OnlineBackup': 'No',
 'DeviceProtection': 'No',
 'TechSupport': 'No',
 'StreamingTV': 'Yes',
 'StreamingMovies': 'No',
 'Contract': 'Month-to-month',
 'PaymentMethod': 'Electronic check',
 'tenure': 1,
 'MonthlyCharges': 59.85,
 'SeniorCitizen': 0,
 'Partner': 'No',
 'Dependents': 'No',
 'PhoneService': 'Yes',
 'PaperlessBilling': 'Yes'}

In [ ]:
# from pprint import pprint
# # формируем запрос
# def send_json(x):
#     pprint(x)
#     body = x    
#     myurl = 'http://6a55-35-233-210-113.ngrok.io' + '/predict'
#     headers = {'content-type': 'application/json; charset=utf-8'}
#     response = requests.post(myurl, json=body, headers=headers)
#     return response.json()['predictions']

In [ ]:
# формируем запрос
def send_json(x):
    gender, multiple_lines, internet_service, online_security, online_backup, device_protection, tech_support, streaming_tv, streaming_movies, contract, payment_method, tenure, monthly_charges, senior_citizen, partner, dependents, phone_service, paperless_billing  = x
# gender               object
# MultipleLines        object
# InternetService      object
# OnlineSecurity       object
# OnlineBackup         object
# DeviceProtection     object
# TechSupport          object
# StreamingTV          object
# StreamingMovies      object
# Contract             object
# PaymentMethod        object
# PhoneService         object
# PaperlessBilling     object
# tenure                int64
# MonthlyCharges      float64
# SeniorCitizen         int64
# Partner              object
# Dependents           object
# dtype: object
    body = {'gender': str(gender),
            'MultipleLines': str(multiple_lines),
            'InternetService': str(internet_service),
            'OnlineSecurity': str(online_security),
            'OnlineBackup': str(online_backup),
            'DeviceProtection': str(device_protection),
            'TechSupport': str(tech_support),
            'StreamingTV': str(streaming_tv),
            'StreamingMovies': str(streaming_movies),
            'Contract': str(contract),
            'PaymentMethod': str(payment_method),
            'tenure': int(tenure),
            'MonthlyCharges': float(monthly_charges),
            'SeniorCitizen': int(senior_citizen),
            'Partner': str(partner),
            'Dependents': str(dependents),
            'PhoneService': str(phone_service),
            'PaperlessBilling': str(paperless_billing),
            }
    myurl = 'http://24d8-35-233-210-113.ngrok.io' + '/predict'
    headers = {'content-type': 'application/json; charset=utf-8'}
    response = requests.post(myurl, json=body, headers=headers)
    return response.json()['predictions']

In [ ]:
# Пример данных
data  = ('Male',
'No',
'DSL',
'No',
'No',
'No',
'No',
'Yes',
'No',
'Month-to-month',
'Electronic check',
1,
59.85,
0,
'No',
'No',
'Yes',
'Yes')

In [ ]:
# обращение к серверу с запросом из одного набора (его построили руками выше - data)
response = send_json(data)
print('предсказание', response)

предсказание 0.26


Сделаем обработку массы запросов:

загрузим данные X_test
https://drive.google.com/file/d/1ztc0UjpO_w-IIfyi2qKdc2cLJaw72P5w/view?usp=sharing

y_test
https://drive.google.com/file/d/1H7UABuFhTyAPuoq9hoZL5ZJ1TE5wAtVG/view?usp=sharing

In [ ]:
!wget 'https://drive.google.com/uc?export=download&id=1ztc0UjpO_w-IIfyi2qKdc2cLJaw72P5w' -O X_test.csv
!wget 'https://drive.google.com/uc?export=download&id=1H7UABuFhTyAPuoq9hoZL5ZJ1TE5wAtVG' -O y_test.csv

--2023-01-25 10:48:21--  https://drive.google.com/uc?export=download&id=1ztc0UjpO_w-IIfyi2qKdc2cLJaw72P5w
Resolving drive.google.com (drive.google.com)... 108.177.12.139, 108.177.12.138, 108.177.12.101, ...
Connecting to drive.google.com (drive.google.com)|108.177.12.139|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-10-bs-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/pgv1kf2ks705cjj5h7d7b6g23cflkr2i/1674643650000/04420567413078840451/*/1ztc0UjpO_w-IIfyi2qKdc2cLJaw72P5w?e=download&uuid=783421ce-4200-4deb-a6a9-9b29b6957df1 [following]
--2023-01-25 10:48:21--  https://doc-10-bs-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/pgv1kf2ks705cjj5h7d7b6g23cflkr2i/1674643650000/04420567413078840451/*/1ztc0UjpO_w-IIfyi2qKdc2cLJaw72P5w?e=download&uuid=783421ce-4200-4deb-a6a9-9b29b6957df1
Resolving doc-10-bs-docs.googleusercontent.com (doc-10-bs-docs.googleusercontent.com)... 173.194.211.132, 2607

In [ ]:
import pandas as pd
from sklearn.metrics import roc_auc_score, roc_curve, precision_recall_curve
from sklearn.metrics import f1_score
from urllib import request, parse


X_test = pd.read_csv("X_test.csv", sep = ',')
y_test = pd.read_csv("y_test.csv", sep = ',')

In [ ]:
categorical_columns = ['gender', 'MultipleLines', 'InternetService', 'OnlineSecurity',  'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 
                       'StreamingMovies', 'Contract', 'PaymentMethod', 'PhoneService', 'PaperlessBilling', 'SeniorCitizen', 'Partner', 'Dependents']
continuous_columns = ['tenure', 'MonthlyCharges', ]
# binary_columns = ['SeniorCitizen', 'Partner', 'Dependents']

features = categorical_columns + continuous_columns
features

['gender',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaymentMethod',
 'PhoneService',
 'PaperlessBilling',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'MonthlyCharges',
 'SeniorCitizen',
 'Partner',
 'Dependents']

In [ ]:
X_test.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  3877-JRJIP    Male              0      No         No       1          Yes   
1  5025-GOOKI  Female              0      No         No      20          Yes   
2  9894-EZEWG  Female              0      No         No       1          Yes   
3  2672-OJQZP  Female              0      No         No       1          Yes   
4  3733-UOCWF    Male              1     Yes         No      61          Yes   
5  9444-JTXHZ    Male              0     Yes         No       1          Yes   
6  5955-EPOAZ  Female              0      No         No       6          Yes   
7  2988-GBIVW  Female              1     Yes         No      70          Yes   
8  4817-QRJSX  Female              0      No         No      10          Yes   
9  9124-LHCJQ  Female              0      No         No       1          Yes   

  MultipleLines InternetService       OnlineSecurity         OnlineBackup  \
0           Yes             DSL                   No                   No   
1            No              No  No internet service  No internet service   
2            No     Fiber optic                   No                  Yes   
3            No             DSL                   No                   No   
4           Yes     Fiber optic                   No                  Yes   
5            No     Fiber optic                   No                   No   
6            No              No  No internet service  No internet service   
7           Yes     Fiber optic                  Yes                   No   
8           Yes     Fiber optic                   No                  Yes   
9           Yes     Fiber optic                   No                   No   

      DeviceProtection          TechSupport          StreamingTV  \
0                   No                   No                  Yes   
1  No internet service  No internet service  No internet service   
2                   No                   No                   No   
3                   No                   No                   No   
4                   No                  Yes                   No   
5                  Yes                   No                   No   
6  No internet service  No internet service  No internet service   
7                   No                   No                   No   
8                   No                   No                   No   
9                   No                   No                   No   

       StreamingMovies        Contract PaperlessBilling  \
0                   No  Month-to-month              Yes   
1  No internet service  Month-to-month              Yes   
2                   No  Month-to-month              Yes   
3                   No  Month-to-month               No   
4                   No        One year              Yes   
5                   No  Month-to-month               No   
6  No internet service  Month-to-month              Yes   
7                  Yes  Month-to-month              Yes   
8                   No  Month-to-month              Yes   
9                  Yes  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges  
0           Electronic check           59.85        59.85  
1    Credit card (automatic)           18.90       347.65  
2    Credit card (automatic)           74.30         74.3  
3               Mailed check           44.75        44.75  
4  Bank transfer (automatic)           85.55      5251.75  
5           Electronic check           76.20         76.2  
6               Mailed check           20.95        109.5  
7           Electronic check           90.85       6470.1  
8    Credit card (automatic)           81.00       818.05  
9               Mailed check           85.45        85.45

In [ ]:
response = send_json(X_test[['gender',
            'MultipleLines',
            'InternetService',
            'OnlineSecurity',
            'OnlineBackup',
            'DeviceProtection',
            'TechSupport',
            'StreamingTV',
            'StreamingMovies',
            'Contract',
            'PaymentMethod',
            'tenure',
            'MonthlyCharges',
            'SeniorCitizen',
            'Partner',
            'Dependents',
            'PhoneService',
            'PaperlessBilling']].iloc[0,:])

In [ ]:
print(response)

0.27


In [ ]:
response = send_json(X_test[['gender',
            'MultipleLines',
            'InternetService',
            'OnlineSecurity',
            'OnlineBackup',
            'DeviceProtection',
            'TechSupport',
            'StreamingTV',
            'StreamingMovies',
            'Contract',
            'PaymentMethod',
            'tenure',
            'MonthlyCharges',
            'SeniorCitizen',
            'Partner',
            'Dependents',
            'PhoneService',
            'PaperlessBilling']].iloc[5,:])

Сделаем N запросов и оценим время

In [ ]:
N = 50

In [ ]:
for i in range(N):

  response = send_json(X_test[['gender',
              'MultipleLines',
              'InternetService',
              'OnlineSecurity',
              'OnlineBackup',
              'DeviceProtection',
              'TechSupport',
              'StreamingTV',
              'StreamingMovies',
              'Contract',
              'PaymentMethod',
              'tenure',
              'MonthlyCharges',
              'SeniorCitizen',
              'Partner',
              'Dependents',
              'PhoneService',
              'PaperlessBilling']].iloc[i,:])
  print(response)

0.27
0.13
0.89
0.76
0.02
0.84
0.14
0.34
0.93
1.0
0.78
0.29
0.775
1.0
0.92
0.1
0.95
0.1
0.03
0.02
0.07
0.04
0.89
1.0
0.0
0.02
0.8466666666666665
0.0
0.49
0.63
0.03
0.37
0.26
0.84
0.02
0.2
0.98
0.21
0.1
0.37
0.84
0.83
0.39
0.11
0.39
0.75
0.0
1.0
1.0
0.22


In [ ]:
%%time
predictions = X_test[['gender',
            'MultipleLines',
            'InternetService',
            'OnlineSecurity',
            'OnlineBackup',
            'DeviceProtection',
            'TechSupport',
            'StreamingTV',
            'StreamingMovies',
            'Contract',
            'PaymentMethod',
            'tenure',
            'MonthlyCharges',
            'SeniorCitizen',
            'Partner',
            'Dependents',
            'PhoneService',
            'PaperlessBilling']].iloc[:N].apply(lambda x: send_json(x), axis=1)

CPU times: user 224 ms, sys: 20.5 ms, total: 245 ms
Wall time: 11.2 s


In [ ]:
predictions.values[:5]

array([0.27, 0.13, 0.89, 0.76, 0.02])

Сделаем N запросов и оценим время

Посчитаем метрику

In [ ]:
y_test.head()

Churn
0      1
1      0
2      1
3      0
4      0

In [ ]:
import numpy as np

precision, recall, thresholds = precision_recall_curve(y_test[:N], predictions)

fscore = (2 * precision * recall) / (precision + recall)
# locate the index of the largest f score
ix = np.argmax(fscore)
print(f'Best Threshold={thresholds[ix]}, F-Score={fscore[ix]:.3f}, Precision={precision[ix]:.3f}, Recall={recall[ix]:.3f}')

Best Threshold=0.775, F-Score=0.872, Precision=0.944, Recall=0.810


In [ ]:
roc_auc_score(y_score=predictions.values, y_true=y_test.values[:N])

0.9556650246305419

Выводы:
На 50 наблюдениях модель показывает неплохие результаты.

Best Threshold=0.775, F-Score=0.872, Precision=0.944, Recall=0.810

Roc_auc_score = 0.9556650246305419
